File: BT/RainCatcher/RainCatcher.java
import java.util.*;

class Solution {
    public int solve(int[] nums) {
        if (nums.length == 0) return 0;

        int[] maxHeightRightToLeft = new int[nums.length];
        int maxHeight = nums[nums.length-1];

        for (int i = nums.length-1; i >= 0; i--) {
            maxHeight = Math.max(maxHeight, nums[i]);
            maxHeightRightToLeft[i] = maxHeight;
        }

        int sum = 0;
        int maxHeightLeftToRight = nums[0];

        for (int i = 0; i < nums.length; i++) {
            maxHeightLeftToRight = Math.max(nums[i], maxHeightLeftToRight);
            sum += Math.min(maxHeightLeftToRight, maxHeightRightToLeft[i]) - nums[i];
        }

        return sum;
    }
}

File: BT/RainCatcher/RainCatcher.py
class Solution:
    def solve(self, nums):
        if not nums:
            return 0
        
        max_heights = [0] * len(nums)
        right_max_height = nums[-1]
        for i in range(len(nums)-1,-1,-1):
            right_max_height = max(right_max_height, nums[i])
            max_heights[i] = max(right_max_height, nums[i])
        
        area = 0
        left_max_height = nums[0]
        for i in range(1, len(nums)):
            left_max_height = max(left_max_height, nums[i])
            area += (min(left_max_height, max_heights[i]) - nums[i])
        return area
        
